src/FlatSharp.Runtime/IO/SpanWriterExtensions.cs

```diff
@@ -36,6 +38,27 @@ public static class SpanWriterExtensions
 
         memory.Span.CopyTo(span.Slice(vectorStartOffset + sizeof(uint)));
     }
+    
+    public static void UnsafeWriteSpan<TSpanWriter, TElement>(
+        this TSpanWriter spanWriter,
+        Span<byte> span,
+        Span<TElement> buffer,
+        int offset,
+        SerializationContext ctx) where TSpanWriter : ISpanWriter where TElement : struct
+    {
+        int numberOfItems = buffer.Length;
+        int vectorStartOffset = ctx.AllocateVector(
+            itemAlignment: Unsafe.SizeOf<TElement>(),
```

Comment:
In FlatBuffers, structs are aligned to according to the max alignment of all their members. So a struct with an int64 member would have 8-byte alignment, etc. This is an artifact of the time when FlatBuffers was designed; some architectures at the time didn't support unaligned reads, so care was taken to align all the data. This turns out to be not a good choice for size on the wire, but the format is the format.

I think the case where a memory copy will be enough to properly serialize a struct is actually pretty narrow: There is no padding between members (that is -- `StructSize % Alignment == 0`). An example would be the typical `Vector3` struct with 3 float32's inside.

However, there are other cases where a memory copy won't be enough:

- Big endian machines. FlatBuffers encodes all data as little endian. FlatSharp at least has code sprinkled throughout trying to be thoughtful about BE architectures. Not that I have a BE machine to test it on, of course, But it's there. And it might even work :)

- Cases where consecutive structs need padding between items (ie, not aligned according to the above).

The fallback for both of these cases would be to use a loop and serialize/parse "normally". You can take a look at what I do for value-type structs [here](https://github.com/jamescourtney/FlatSharp/blob/2ec1312212655f5bf05e9678f3ec205aad3a410e/src/FlatSharp/TypeModel/ValueStructTypeModel.cs#L171). Basically, the logic is that if memory-marshalling is allowed by the user and it's a LE machine, then do it. Otherwise, go field-by-field.

That said, if Unity is LE only/mostly and you think it would be better to raise the error about poor alignment rather than have people get unexpectedly bad performance, then raising an error might be reasonable.

Reply:
For some examples, I have a few "nightmare struct" tests:

https://github.com/jamescourtney/FlatSharp/blob/2ec1312212655f5bf05e9678f3ec205aad3a410e/src/Tests/FlatSharpTests/SerializationTests/VectorSerializationTests.cs#L225
https://github.com/jamescourtney/FlatSharp/blob/2ec1312212655f5bf05e9678f3ec205aad3a410e/src/Tests/FlatSharpTests/SerializationTests/VectorSerializationTests.cs#L307

Reply:
Finally, I think I might also suggest experimenting to see if this should be inlined or not. Does IL2CPP actually honor that attribute?